cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go

```diff
@@ -95,130 +95,123 @@ func (o *checkCapacityProvClass) Provision(
-func (o *checkCapacityProvClass) checkcapacity(unschedulablePods []*apiv1.Pod, provReq *provreqwrapper.ProvisioningRequest) (bool, error) {
-	var capacityAvailable bool
-	err, cleanupErr := clustersnapshot.WithForkedSnapshot(o.context.ClusterSnapshot, func() (bool, error) {
-		st, _, err := o.schedulingSimulator.TrySchedulePods(o.context.ClusterSnapshot, unschedulablePods, scheduling.ScheduleAnywhere, true)
-		if len(st) < len(unschedulablePods) || err != nil {
-			if noRetry, ok := provReq.Spec.Parameters[NoRetryParameterKey]; ok && noRetry == "true" {
-				// Failed=true condition triggers retry in Kueue. Otherwise ProvisioningRequest with Provisioned=Failed
-				// condition block capacity in Kueue even if it's in the middle of backoff waiting time.
-				conditions.AddOrUpdateCondition(provReq, v1.Failed, metav1.ConditionTrue, conditions.CapacityIsNotFoundReason, "CA could not find requested capacity", metav1.Now())
-			} else {
-				if noRetry, ok := provReq.Spec.Parameters[NoRetryParameterKey]; ok && noRetry != "false" {
-					klog.Errorf("Ignoring Parameter %v with invalid value: %v in ProvisioningRequest: %v. Supported values are: \"true\", \"false\"", NoRetryParameterKey, noRetry, provReq.Name)
-				}
-				conditions.AddOrUpdateCondition(provReq, v1.Provisioned, metav1.ConditionFalse, conditions.CapacityIsNotFoundReason, "Capacity is not found, CA will try to find it later.", metav1.Now())
-			}
-			capacityAvailable = false
-		} else {
-			conditions.AddOrUpdateCondition(provReq, v1.Provisioned, metav1.ConditionTrue, conditions.CapacityIsFoundReason, conditions.CapacityIsFoundMsg, metav1.Now())
-			capacityAvailable = true
-		}
+func (o *checkCapacityProvClass) isBatchEnabled() bool {
+	return o.provreqInjector != nil && o.checkCapacityProvisioningRequestMaxBatchSize > 1
+}
 
-		_, updErr := o.client.UpdateProvisioningRequest(provReq.ProvisioningRequest)
-		if updErr != nil {
-			return false, fmt.Errorf("failed to update Provisioned condition to ProvReq %s/%s, err: %v", provReq.Namespace, provReq.Name, updErr)
+func (o *checkCapacityProvClass) checkCapacityBatch(reqs []provreq.ProvisioningRequestWithPods, combinedStatus *combinedStatusSet, startTime time.Time) []*provreqwrapper.ProvisioningRequest {
+	updates := make([]*provreqwrapper.ProvisioningRequest, 0, len(reqs))
+	for _, req := range reqs {
+		// timebox checkCapacity when batch processing.
+		if o.isBatchEnabled() && time.Since(startTime) > o.checkCapacityProvisioningRequestBatchTimebox {
+			klog.Infof("Batch timebox exceeded, processed %d check capacity provisioning requests this iteration", len(updates))
+			break
 		}
 
-		return capacityAvailable, nil
-	})
-
-	if cleanupErr != nil {
-		return false, cleanupErr
-	}
+		if err := o.checkCapacity(req.Pods, req.PrWrapper, combinedStatus); err != nil {
+			klog.Errorf("error checking capacity %v", err)
+			continue
```

Comment:
Couldn't we keep at least one error if it happens and return it after processing the batch? It could be useful if all calls to checkCapacity fail.

Reply:
Differences:
1. Returning AutoscalerError from loop will register the loop as 'unsuccessful' in a different way than 'nothing-to-do-loop'. If CA keeps running into errors every loop for some time (I think it's 15m by default or sth), it'll start failing healthcheck and the liveness probe will restart it.
2. Possibly error reporting to the user (if/when we do any for ProvisioningRequests)

I think this change can be done in a follow-up PR, if we decide it's necessary.